bog.el

```diff
@@ -1181,25 +1183,32 @@ level `bog-refile-maxlevel' are considered."
     (cdr (assoc-string (completing-read "File: " note-paths)
                        note-paths))))
 
+(defvar bog--agenda-map
+  (let ((map (make-sparse-keymap)))
+    (set-keymap-parent map org-agenda-mode-map)
+    (define-key map "r" 'bog-agenda-redo)
+    (define-key map "g" 'bog-agenda-redo)
+    map)
+  "Local keymap for Bog-related agendas.")
+
 (defmacro bog--with-search-lprops (&rest body)
   "Execute BODY with Bog-related agenda values.
 Restore the `org-lprops' property value for
 `org-agenda-redo-command' after executing BODY."
-  (declare (indent 0))
-  `(let ((org-lprops (get 'org-agenda-redo-command 'org-lprops))
-         (bog-lprops '((org-agenda-buffer-name "*Bog search*")
-                       (org-agenda-files (bog-notes))
-                       org-agenda-text-search-extra-files
-                       org-agenda-sticky)))
-     (put 'org-agenda-redo-command 'org-lprops bog-lprops)
-     (put 'org-agenda-files 'org-restrict nil)
-     (org-let bog-lprops ,@body)
-     (use-local-map (let ((map (make-sparse-keymap)))
-                      (set-keymap-parent map org-agenda-mode-map)
-                      (define-key map "r" 'bog-agenda-redo)
-                      (define-key map "g" 'bog-agenda-redo)
-                      map))
-     (put 'org-agenda-redo-command 'org-lprops org-lprops)))
+  (declare (indent 0) (debug t))
+  (let ((bog-lprops '((org-agenda-buffer-name "*Bog search*")
+                      (org-agenda-files (bog-notes))
+                      org-agenda-text-search-extra-files
+                      org-agenda-sticky))
+        (org-lprops (make-symbol "org-lprops")))
+    `(let ((,org-lprops (get 'org-agenda-redo-command 'org-lprops)))
+       (unwind-protect
+           (let ,bog-lprops
+             (put 'org-agenda-redo-command 'org-lprops ',bog-lprops)
+             (put 'org-agenda-files 'org-restrict nil)
+             ,@body
+             (use-local-map bog--agenda-map))
+         (put 'org-agenda-redo-command 'org-lprops ,org-lprops)))))
```

Comment:
BTW I think this can also be written as follows, if you prefer:

```el
(let ((bog-lprops '((org-agenda-buffer-name "*Bog search*")
                    (org-agenda-files (bog-notes))
                    (org-agenda-text-search-extra-files ())
                    (org-agenda-sticky nil))))
  `(cl-letf (((get 'org-agenda-redo-command 'org-lprops) ',bog-lprops)
             ,@bog-lprops)
     (put 'org-agenda-files 'org-restrict nil)
     ,@body
     (use-local-map bog--agenda-map)))
```

Reply:
I thought I recalled trying to switch over to lexical binding.  Checking my local repo, I have a branch from 2015 that basically says "didn't get agenda stuff working".

Looks like you managed to get it working nicely.  I did a good amount of manual testing around `bog-search{,-for-citekey}`, as well as the interaction with the regular agenda, and didn't spot any issues.

I also tested with the version you proposed in the comments and that also works well.  Whichever you prefer is fine by me.

Reply:
Thanks for testing and confirming. I tend to judiciously avoid overusing `cl-lib`, but in this case it probably conveys better what the code is trying to do, so I switched to that version.